### DISCUSSION TITLE
vsg::RegionOfInterest node checked in, useful for guiding extents of shadows etc.

### DISCUSSION THREAD

**Discussion by robertosfield** [vsg-dev/VulkanSceneGraph#1146] [General]
Hi All,

Ever since I started work on shadows in the VSG I've been contemplating a mechanism that could allow users to specify what extents of the scene graph to use when computing the shadow extents.  Today I added the [vsg::RegionOfInteret](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/nodes/RegionOfInterest.h#L21) node for this purpose, and added support for it into VSG, now merged master, and added to the vsgshadow example to test it out.

* [Changes to VulkanScenegraph](https://github.com/vsg-dev/VulkanSceneGraph/pull/1145)
* [Changes to vsgExamples](https://github.com/vsg-dev/vsgExamples/pull/302)

The vsg::RegionOfInterest is a node that contains a list of dvec3 points that encompass the region of interest, optimally set up this would be the points of a convex hull thought this optimal configuration isn't required.  You can place multiple RegionOfInterest nodes in your scene graph and nested them underneath transform nodes etc. 

The vsgshadow additions illustrate how you can set up  vsg::RegionOfInterest using the corners of the compute bound of a subgraph:

~~~ cpp
    if (assignRegionOfInterest)
    {
        auto group = vsg::Group::create();

        auto regionOfInterest = vsg::RegionOfInterest::create();
        regionOfInterest->name = "region to cast shadow";
        regionOfInterest->points.emplace_back(bounds.min.x, bounds.min.y, bounds.min.z);
        regionOfInterest->points.emplace_back(bounds.max.x, bounds.min.y, bounds.min.z);
        regionOfInterest->points.emplace_back(bounds.max.x, bounds.max.y, bounds.min.z);
        regionOfInterest->points.emplace_back(bounds.min.x, bounds.max.y, bounds.min.z);
        regionOfInterest->points.emplace_back(bounds.min.x, bounds.min.y, bounds.max.z);
        regionOfInterest->points.emplace_back(bounds.max.x, bounds.min.y, bounds.max.z);
        regionOfInterest->points.emplace_back(bounds.max.x, bounds.max.y, bounds.max.z);
        regionOfInterest->points.emplace_back(bounds.min.x, bounds.max.y, bounds.max.z);
        group->addChild(regionOfInterest);
        group->addChild(scene);

        scene = group;
    }
~~~

The vsg::RecordTraversal now tracks the vsg::RegionOfInterest nodes it comes across in a RecordTraversal::regionsOfInterest list and the  vsg::ViewDependentState class that handles the shadow map creation now uses this list, transforms the points they enclose into eye coordinates and then adjusts the effective near/far when setting up the extents of the light space.  What this means is the shadow maps are created for a more tightly bound region which increases the effective resolution of the generated shadow maps.

This is the results without RegionOfInterest node:
~~~ sh 
vsgshadow --large 
~~~
![Screenshot_20240403_131307](https://github.com/vsg-dev/VulkanSceneGraph/assets/1327952/7805b817-50ac-4ebc-bd02-9c1b1a9beaa8)

With RegionOfInterest node:
~~~ sh 
vsgshadow --large --region-of-interest
~~~
![Screenshot_20240403_131548](https://github.com/vsg-dev/VulkanSceneGraph/assets/1327952/52ded8af-1dea-4bb8-beb5-6fe9766508c4)

The vsg::RegionOfInterest node also has a mask member to enable different views to select different combinations of RegionOfInterest that should be enabled, and name member to aid to organization of different regions of interest.

While I have specifically implemented RegionOfInterest support into the shadow mapping backend I have kept the new node decoupled from the specifics of lighting/shadows, it's a general purpose class that you can use for your own purposes and potentially we can use it other places of the VSG such as clamping the projection matrix near/far to the extents of RegionOfInterest.

**Comment by hytel** [comment]
Looking through the vsgshadow code, it only adds **one region** if you turn on with --region-of-interest, and it's for the whole scene bounds.  But the scene bounds here is only 1000x1000.  In my code, my scene is much larger, but way fewer objects.  Say 50,000 x 50,000 and two objects.  But each object is near opposite corners.  When they are far away from each other, the PCSS shadows are **very** granular, and much better when they are close.  I made an ROI for each object, but it doesn't really help the render quality.  If I point the camera so that it can see the close one, but it's pointing away from the one far away the shadows are sharp.  Is there a way to render nice soft shadows for 1-64 objects spread out over a larger scene sizes for any camera look direction?

**Reply by ArepMM** [reply]
In our project after some brainstorm we built shadow region as 5 points of camera's view pyramid:
<img width="640" height="480" alt="image" src="https://github.com/user-attachments/assets/1dcfd0ad-a6b0-4c68-b640-b64dd795c829" />
And then update it every frame to follow camera moving. Implementation example:

#include <vsg/core/Visitor.h>
#include <vsg/app/Camera.h>
#include <vsg/nodes/RegionOfInterest.h>

class UpdateShadowRegionHandler final : public vsg::Inherit<vsg::Visitor, UpdateShadowRegionHandler>
{
public:
    
    vsg::ref_ptr<vsg::Camera> camera;
    vsg::ref_ptr<vsg::RegionOfInterest> shadow_region;
    double shadows_draw_distance;
    
    UpdateShadowRegionHandler(vsg::ref_ptr<vsg::Camera> in_camera,
                              vsg::ref_ptr<vsg::RegionOfInterest> in_shadow_region,
                              double in_shadows_draw_distance = 50.0)
        : Inherit()
        , camera(in_camera)
        , shadow_region(in_shadow_region)
        , shadows_draw_distance(in_shadows_draw_distance)
    {
        shadow_region->points.resize(5);
    };
    
    void apply(vsg::FrameEvent& frame) override
    {
        vsg::ref_ptr<vsg::LookAt> lookAt = camera->viewMatrix.cast<vsg::LookAt>();
        
        vsg::dvec3 view_orth = vsg::normalize(lookAt->center - lookAt->eye);
        vsg::dvec3 view_up = vsg::normalize(lookAt->up);
        vsg::dvec3 view_right = vsg::cross(view_orth, view_up);
        
        vsg::dvec3 point_center = lookAt->eye + view_orth * shadows_draw_distance;
        vsg::dvec3 vec_up = view_up * (shadows_draw_distance / 2.0);
        vsg::dvec3 vec_right = view_right * (shadows_draw_distance / 2.0);
        
        shadow_region->points[0] = lookAt->eye;
        shadow_region->points[1] = point_center + vec_up + vec_right;
        shadow_region->points[2] = point_center + vec_up - vec_right;
        shadow_region->points[3] = point_center - vec_up + vec_right;
        shadow_region->points[4] = point_center - vec_up - vec_right;
    }
};

**Reply by hytel** [reply]
I'm not using the vsg::LookAt view matrix (using a custom one), but the idea can easily be used.  Thanks, this is perfect!
